### PR TITLE
Evaluator accumulate

### PR DESCRIPTION
**Pull request by typhoonzero**
Fix https://github.com/PaddlePaddle/Paddle/issues/3808

### CODE REVIEW

File: python/paddle/v2/framework/tests/test_evaluator.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+from paddle.v2.framework.op import Operator
+import paddle.v2.framework.core as core
+import unittest
+import op_test
+import numpy as np
+
+
+class TestEvaluator(unittest.TestCase):
+    def setup(self, scope, inputs, outputs):
+        def __create_var__(var_name, arr):
```
Comment (qingqing01) [inline python/paddle/v2/framework/tests/test_evaluator.py]: `def __create_var__` -> `def _create_var`.

File: python/paddle/v2/framework/evaluator.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+def avg_accumulate(accumulated_var, per_eval, num_batches, place):
+    t = np.array(accumulated_var.get_tensor())
+    t[0] += per_eval[0]
+    accumulated_var.get_tensor().set([t[0] / float(num_batches)], place)
+
+
+class Evaluator(object):
+    def __init__(self,
+                 scope,
+                 operator='accuracy',
```
Comment (qingqing01) [inline python/paddle/v2/framework/evaluator.py]: 看了下这个`Evaluator`里只有对多个mini-batch间的结果累加求平均。但对于PrecisionRecall这种多个mini-batch间结果统计并不是简单的累加求平均。

所以感觉如果在Python里抽象出`Evaluator`，用来统计多个mini-batch间的计算结果，需要抽象出一个`Evaluator`基类接口，然后有很多子类，AccuracyEvalutor, PrecisionRecallEvalutor等等。

**Comment by helinwang** [comment]
@qingqing01 @typhoonzero @wangkuiyi from my understanding the PaddlePaddle program is divided into [compile time (currently Python) and runtime (C++)](https://github.com/PaddlePaddle/Paddle/blob/4838ea25d3cdb387cda0364e06251de1095bc1c6/doc/design/var_desc.md#background): the Python code generates a protobuf which describes the computation, and the C++ code runs the computation description.

This implementation of the executor is in Python, I am not sure if it fits with the compile time and runtime concept. I think this requirement is very important for every OP implementation, @qingqing01 @wangkuiyi perhaps we need to make sure that everyone is on the same page.

**Comment by typhoonzero** [comment]
@helinwang you are right, I'll try to put this implementation to compile time.

File: python/paddle/v2/framework/evaluator.py (hunk 2 above)
Comment (typhoonzero) [reply python/paddle/v2/framework/evaluator.py]: 嗯，前面定义了`avg_accumulate`，是解决这个问题的。但的确没有考虑编译时。需要把avg accumulator等作为op，才可以使python定义的计算是编译时的。

File: python/paddle/v2/framework/evaluator.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        t.set_dims((1, ))
+        t.set([0.0], place)
+        # self.accumulated_var = block.create_var(block, name=eval_name, shape=(1,))
+        # self.accumulated_var.get_tensor().set([0.0])
+        # create operator of evaluation
+        var_map = dict()  # var name -> variable
+        var_map[input] = [input]
+        var_map[label] = [label]
+        var_map[output] = [output]
+        self.op = op.Operator(operator, **var_map)
```
Comment (dzhwinter) [inline python/paddle/v2/framework/evaluator.py]: Currently, we can not create operator immediately out of Python `Block` and `Program`.
https://github.com/PaddlePaddle/Paddle/blob/1f11f773bf761171288b165984bc26a379fe1db8/python/paddle/v2/framework/framework.py#L430-L446  

May I just merge this work into codebase and let me polish it to fit in the refactorization code? : )
Our book porting work need this feature eagerly. @typhoonzero @qingqing01

Comment (typhoonzero) [reply python/paddle/v2/framework/evaluator.py]: Sure, Thanks. Please also refer to some of my thoughts: https://github.com/PaddlePaddle/Paddle/issues/5157

**Pull request merged**